src/openforms/formio/dynamic_config/date.py

```diff
@@ -0,0 +1,143 @@
+    minDate: Optional[DateConstraintConfiguration]
+    maxDate: Optional[DateConstraintConfiguration]
+
+
+class DatePickerConfig(TypedDict):
+    minDate: Optional[str]
+    maxDate: Optional[str]
+
+
+class FormioDateComponent(Component):
+    openForms: Optional[OpenFormsConfig]
+    datePicker: Optional[DatePickerConfig]
+
+
+DatetimeLike = Optional[Union[datetime, str]]
+
+
+@register("date")
+class DateComponent(BasePlugin):
+    """
+    Implement behaviour for our custom date component
+
+    In the JS, this component type inherits from Formio datetime component. See
+    ``src/openforms/js/components/form/date.js`` for the various configurable options.
+    """
+
+    verbose_name = _("Date component")
+
+    @classmethod
+    def mutate(cls, component: FormioDateComponent, data: DataMapping) -> None:
+        for key in ("minDate", "maxDate"):
+            config = cast(
+                DateConstraintConfiguration,
+                glom(component, f"openForms.{key}", default=None),
+            )
+            if config is None:
+                continue
+
+            if (mode := config.get("mode")) is None:
+                continue
+
+            # formio has set the datePicker.{key} value
+            if mode == "fixedValue":
+                continue
+
+            config = cls.normalize_config(config)
+            value = cls.calculate_delta(config, data)
+            if value and isinstance(value, datetime):
+                value = value.isoformat()
+            assign(component, f"datePicker.{key}", value, missing=dict)
+
+    @staticmethod
+    def normalize_config(
+        config: DateConstraintConfiguration,
+    ) -> DateConstraintConfiguration:
+        mode = config["mode"]
+        assert mode in ("future", "past", "relativeToVariable")
+        if mode == "relativeToVariable":
+            return config
+
+        # mode is now future or past -> convert that to a relative delta config
+        config["mode"] = "relativeToVariable"
+        config["variable"] = NOW_VARIABLE
+        include_today = cast(bool, glom(config, "includeToday", default=False))
+        config["operator"] = "add" if mode == "future" else "subtract"
+
+        delta = cast(
+            DateConstraintDelta,
+            {
+                "years": 0,
+                "months": 0,
+                "days": 0 if include_today else 1,
+            },
+        )
+        config["delta"] = delta
+        return config
+
+    @staticmethod
+    def calculate_delta(
+        config: DateConstraintConfiguration,
+        data: DataMapping,
+    ) -> DatetimeLike:
+        assert config["mode"] == "relativeToVariable"
+
+        base_value = cast(
+            Optional[datetime],
+            glom(data, config["variable"], default=None),
+        )
+        # can't do calculations on values that don't exist
+        if base_value is None:
+            return None
+
+        assert (
+            base_value.tzinfo is not None
```

Comment:
But, date and time components (and the values people put in) are not time aware. Do we/formio convert them somewhere else?

Reply:
Our `date` component inherits from Formio's `datetime` component (which out of the box has options to disable the date or time aspect), and that's why it internally stores and communicates everything using full datetime strings.

Dates also inherently lack timezone information - Sept 13 at 1am midnight is Sept 13 in NL, but Sept 12 in UTC, so you need that timezone information to be able to figure out the correct date in the local timezone of the user.